### PR TITLE
Wallet factory redesign

### PR DESCRIPTION
**Pull request by livingrockrises**
As per the discussion, this change can ease the deployment of SCW on multiple chains while keeping address the same. In other words without needing to track the information like implementation, initialiser(handler address etc) to deploy on another chain.

For any case we would be upgrading to new implementation
i. enforced : batch when user sends first userop, when user sends userop and upgrade is necessary (implementation change due to entry point change etc)
ii. informed: by dapps and giving them ability to just call setImplementation etc

Benefits:

All the wallets will be deployed at the same address for same owner and index(0,1,2..)
Deployed proxy will have accurate owner storage, minimal handler and functional audited secure implementation.

once deployed handler can be set new and implementation can be set new.

below is the case in which wallet deployment and upgrades can be done as part of single userop.

initcode : wallet factory + deployment data  
calldata: executeCall, executeBatchCall [ where batch is: self:updateImplementationAndCall(,), self:setFallbackHandler, self:updateImplementation, dapp:dappmethod ]

WIP.....

updating test cases as updateImplementation() is not requiredFromEntryPointOrOwner anymore, but it is mixedAuth. which means userOp.calldata should not be updateImplementation anymore but executeCall and _to:wallet, _data:updateImplementation function signature.

new entry point comes and implementation must be upgraded. above first deploy and call should go through prior EP and then onwards use new entrypoint!

methods in the sdk for setImplementation and setFallbackHandler etc and prepating userops..

### CODE REVIEW

File: contracts/smart-contract-wallet/SmartAccountFactory.sol [hunk 1]
```diff
@@ -101,6 +115,8 @@ contract SmartAccountFactory {
         }
         require(address(proxy) != address(0), "Create call failed");
 
+        bytes memory initializer = getInitializer(_owner);
```
Comment (tomarsachin2271) [inline contracts/smart-contract-wallet/SmartAccountFactory.sol:118]: Also emit a different event in this function as well.

File: contracts/smart-contract-wallet/utils/GasEstimatorSmartAccount.sol [hunk 2]
```diff
@@ -9,15 +9,14 @@ contract GasEstimatorSmartAccount {
...
-        uint _index,
+        address _owner,
+        uint256 _index,
         bytes calldata _data // execTransaction data // counterFactual wallet should have assets if required
     ) external returns (bool success, bytes memory result, uint256 gas) {
         // solhint-disable
         uint256 initialGas = gasleft();
         address _wallet = SmartAccountFactory(_factory)
-            .deployCounterFactualWallet(_implementation, initializer, _index);
+            .deployCounterFactualWallet(_owner, _index);
```
Comment (tomarsachin2271) [inline contracts/smart-contract-wallet/utils/GasEstimatorSmartAccount.sol:19]: Do we call this method using eth_call?

**Review by tomarsachin2271 — COMMENTED** [review]
Changes looks good. Added 2 comments.

File: contracts/smart-contract-wallet/utils/GasEstimatorSmartAccount.sol (hunk 2 above)
Comment (livingrockrises) [reply contracts/smart-contract-wallet/utils/GasEstimatorSmartAccount.sol:19]: yes, this contract is for gas estimation test cases when wallet is not already deployed..

File: contracts/smart-contract-wallet/SmartAccountFactory.sol (hunk 1 above)
Comment (livingrockrises) [reply contracts/smart-contract-wallet/SmartAccountFactory.sol:118]: what would that event be?

Comment (tomarsachin2271) [reply contracts/smart-contract-wallet/SmartAccountFactory.sol:118]: In deployCounterFactualWallet, emit event `CounterFactualAccountCreated`
and in deployWallet emit event `AccountCreated`

Comment (livingrockrises) [reply contracts/smart-contract-wallet/SmartAccountFactory.sol:118]: I see. ok

**Pull request merged**